p2p/tools/local_network.py

```diff
@@ -0,0 +1,378 @@
+        self.name = name
+        self.router = router
+        self.default_host = default_host
+
+    @property
+    def default_host(self) -> str:
+        if self._default_host is not None:
+            return self._default_host
+        else:
+            return os.environ.get('P2P_DEFAULT_HOST', '127.0.0.1')
+
+    @default_host.setter
+    def default_host(self, value: str) -> None:
+        self._default_host = value
+
+    #
+    # Asyncio API
+    #
+    async def start_server(
+            self,
+            client_connected_cb: ConnectionCallback,
+            host: str,
+            port: int) -> Server:
+        host = self.router.get_host(host)
+        return await host.start_server(client_connected_cb, port)
+
+    async def open_connection(
+            self,
+            host: str,
+            port: int) -> ReaderWriterPair:
+        client_host = self.router.get_host(self.default_host)
+        try:
+            return await client_host.open_connection(host, port)
+        except ConnectionRefusedError as err:
+            # if we fail to connect to the specified host, check if there is a
+            # server running on `0.0.0.0` and connect to that.
+            catch_all_host = self.router.get_host('0.0.0.0')
+            try:
+                return await catch_all_host.open_connection('0.0.0.0', port)
+            except ConnectionRefusedError:
+                pass
+            raise err
+
+
+class Host:
+    servers: Dict[int, Server] = None
+    connections: Dict[int, Address] = None
+
+    def __init__(self, host: str, router: Router) -> None:
+        self.router = router
+        self.host = host
+        self.servers = {}
+        self.connections = {}
+
+    def get_server(self, port: int) -> Server:
+        try:
+            return self.servers[port]
+        except KeyError:
+            raise ConnectionRefusedError("No server running at {0}:{1}".format(self.host, port))
+
+    def _get_open_port(self) -> int:
+        while True:
+            port = random.randint(2**15, 2**16 - 1)
+            if port in self.connections:
+                continue
+            elif port in self.servers:
+                continue
+            else:
+                break
+        return port
+
+    async def start_server(self, client_connected_cb: ConnectionCallback, port: int) -> Server:
+        if port in self.servers:
+            raise OSError('Address already in use')
+
+        address = Address(self.host, port)
+
+        server = Server(client_connected_cb, address)
+        self.servers[port] = server
+        return server
+
+    def receive_connection(self, port: int) -> ReaderWriterPair:
+        address = Address(self.host, port)
+        if port not in self.servers:
+            raise ConnectionRefusedError("No server running at {0}:{1}".format(self.host, port))
+        elif address.port in self.connections:
+            raise OSError('Address already in use')
```

Comment:
Ditto, although I guess you might want to match *exatly* what python does, even though it kinda sucks. I think showing the port is better, but I get either choice.